### PR TITLE
SRCHEIGHT/SRCWIDTH for GetLegendGraphic request

### PR DESCRIPTION
**Pull request by signedav**
I my implementation here #9236 regarding `GetLegendGraphics` I used the `WIDTH` and `HEIGHT` (like already used for content dependent legends) as the width and height of the map that I can determine the size of the symbols with the bounding box. But on a `GetLegendGraphics` request these parameters are used for the size of the generated legend image. Usually. In QGIS Server they are only used for the legend image size, when there is a `RULE` passed. And if there is a `BBOX` it does not allow a combination `RULE`. If no `RULE` defined it takes the minimum size. So everything works fine still with that implementation. 

Still, it would make sense to have a non-standard parameter defining the with and the height used on the map on a `GetLegendGraphic` that is content depending. Since [Geoserver](https://docs.geoserver.org/latest/en/user/services/wms/get_legend_graphic/#content-dependent-legends) introduced `SRCHEIGHT` and `SRCWIDTH` I used the same keywords.

Since there are already tests that use `HEIGHT` and `WIDTH` on `GetLegendGraphic` requests wrongly I assume, that there are also users, that use it wrongly. So I added tha fallback, that if `SRCHEIGHT` is 0 it still tries to take `HEIGHT`.

### CODE REVIEW

**Comment by signedav** [comment]
@pblottiere what do you think about this?

File: src/server/services/wms/qgswmsparameters.cpp [hunk 1]
```diff
@@ -680,6 +690,41 @@ namespace QgsWms
...
+  {
+    return mWmsParameters[ QgsWmsParameter::SRCWIDTH ].toInt();
+  }
+
+  int QgsWmsParameters::getHeightAsInt() const
+  {
+    if ( QStringList( { QStringLiteral( "GetLegendGraphic" ), QStringLiteral( "GetLegendGraphics" ) } ).contains( request() ) && srcHeightAsInt() > 0 )
+      return srcHeightAsInt();
+    return heightAsInt();
+  }
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsparameters.cpp]: I think that `QgsWmsParameters` should not manage requests specificities. Could you do the magic within the `QgsRenderer::getFeatureInfo` instead?

**Comment by pblottiere** [comment]
Hi @signedav,

> Still, it would make sense to have a non-standard parameter defining the with and the height used on the map on a GetLegendGraphic that is content depending. Since Geoserver introduced SRCHEIGHT and SRCWIDTH I used the same keywords.

AFAIAC +1.

@elpaso @rldhont what do you think?

**Comment by elpaso** [comment]
> Hi @signedav,
> 
> > Still, it would make sense to have a non-standard parameter defining the with and the height used on the map on a GetLegendGraphic that is content depending. Since Geoserver introduced SRCHEIGHT and SRCWIDTH I used the same keywords.
> 
> AFAIAC +1.
> 
> @elpaso @rldhont what do you think?

I agree: better don't mix WMS keywords with vendor parameters,  SRCHEIGHT and SRCWIDTH seem the good choice to me.

File: src/server/services/wms/qgswmsparameters.cpp [hunk 2]
```diff
@@ -680,6 +690,41 @@ namespace QgsWms
...
+
+  int QgsWmsParameters::srcWidthAsInt() const
+  {
+    return mWmsParameters[ QgsWmsParameter::SRCWIDTH ].toInt();
+  }
+
+  int QgsWmsParameters::getHeightAsInt() const
+  {
+    if ( QStringList( { QStringLiteral( "GetLegendGraphic" ), QStringLiteral( "GetLegendGraphics" ) } ).contains( request() ) && srcHeightAsInt() > 0 )
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsparameters.cpp]: This looks quite inefficient - either use a static QStringList or just put the two conditions separately

File: src/server/services/wms/qgswmsparameters.h [hunk 3]
```diff
@@ -370,6 +372,50 @@ namespace QgsWms
...
+       * \throws QgsBadRequestException
+       */
+      int srcHeightAsInt() const;
+
+      /**
+       * Returns SRCHEIGHT parameter if it's a GetLegendGraphics request and otherwise HEIGHT parameter
+       *
+       * \returns getHeightAsInt parameter
+       */
+      int getHeightAsInt() const;
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsparameters.h]: Can you come up with a more descriptive name here? The API looks quite confusing when reading the changes below -- ("heightAsInt()" -> "getHeightAsInt()" ) ... looks very likely to be misused.

File: src/server/services/wms/qgswmsparameters.cpp (hunk 2 above)
Comment (pblottiere) [reply src/server/services/wms/qgswmsparameters.cpp]: > This looks quite inefficient - either use a static QStringList or just put the two conditions separately

You're totally right @nyalldawson. But once moved within `QgsRenderer::getLegendGraphics` (cf my previous comment https://github.com/qgis/QGIS/pull/9545#pullrequestreview-215585729), this method (as well as these string comparisons) won't even be necessary (and the misused method's name will be gone too).

Comment (signedav) [reply src/server/services/wms/qgswmsparameters.cpp]: Ok, thanks. I'll come up with an update regarding this soon...

**Comment by signedav** [comment]
@pblottiere sorry for not doing something on that earlier. Is it now like you prefer? I moved the deciding part to `QgsRenderer` but not removed it completely. Why I decide what parameter is used on the `request()` is, because the parameter is requested in several functions, sometimes used by `GetMap` and `GetLegendGraphics`...

**Comment by pblottiere** [comment]
Hi @signedav,

Thanks for your modifications :+1:!

But I think that `getHeightAsInt()` and `getWidthAsInt()` methods are not useful. Actually, it's only needed for `getLegendGraphics` and you can pass the width / height to the `createImage` method directly. In other cases (like `getFeatureInfo` or `scaleImage`), it doesn't seem necessary.

What do you think? Am I missing something?

File: src/server/services/wms/qgswmsparameters.h [hunk 4]
```diff
@@ -390,6 +392,36 @@ namespace QgsWms
        */
       int heightAsInt() const;
 
+      /**
+       * Returns SRCWIDTH parameter or an empty string if not defined.
+       * \returns srcWidth parameter
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsparameters.h]: \since

File: src/server/services/wms/qgswmsparameters.h [hunk 5]
```diff
@@ -390,6 +392,36 @@ namespace QgsWms
...
+       */
+      QString srcWidth() const;
+
+      /**
+       * Returns SRCWIDTH parameter as an int or its default value if not
+       * defined. An exception is raised if SRCWIDTH is defined and cannot be
+       * converted.
+       * \returns srcWidth parameter
+       * \throws QgsBadRequestException
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsparameters.h]: \since

File: src/server/services/wms/qgswmsparameters.h [hunk 6]
```diff
@@ -390,6 +392,36 @@ namespace QgsWms
...
+       */
+      int srcWidthAsInt() const;
+
+      /**
+       * Returns SRCHEIGHT parameter or an empty string if not defined.
+       * \returns srcHeight parameter
+       */
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsparameters.h]: \since

File: src/server/services/wms/qgswmsparameters.h [hunk 7]
```diff
@@ -390,6 +392,36 @@ namespace QgsWms
...
+      QString srcHeight() const;
+
+      /**
+       * Returns SRCHEIGHT parameter as an int or its default value if not
+       * defined. An exception is raised if SRCHEIGHT is defined and cannot be
+       * converted.
+       * \returns srcHeight parameter
+       * \throws QgsBadRequestException
+       */
+      int srcHeightAsInt() const;
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsparameters.h]: \since

**Comment by signedav** [comment]
Thanks for the feedback @pblottiere 

I agree that we can keep the direct WIDTH/HEIGHT request on `GetFeatureInfo`. 

On `scaleImage` and `checkMaximumWidth` the width and height is requested in cases of `getMap` what is used in case of `contentBasedLegend`. As you say, the scale and the maximumWidth is not important at all there and it would work with the image size assumed as the map size as well. But I implemented it like this, that it does not misuse the image size at all.

But in the end I don't have a strong opinion. Probably you're right and it's more confusing to have these methods than that it uses the image sizes for the map (when the size does not matter anyway).

**Comment by signedav** [comment]
I thought about it again and found another critical point if I would remove the functions: on a `GetLegendGraphics` request (in case of `contentBasedLegends`) it calls the `getMap` function here:
https://github.com/qgis/QGIS/blob/44c395ced8ca996d8bb837755207ab0c115823e3/src/server/services/wms/qgswmsrenderer.cpp#L3317
On `getMap` it checks for the maximum size and there it compares it to the value in the parameter `HEIGHT` or `SRCHEIGHT`:
https://github.com/qgis/QGIS/blob/44c395ced8ca996d8bb837755207ab0c115823e3/src/server/services/wms/qgswmsrenderer.cpp#L1997
If we don't decide on the `REQUEST` what `HEIGHT` is used there, it can be, that on a `GetLegendGraphics` request we compare the legend-image-size to the max-size of the map, what could lead to a wrong error message.

Only alternative for a more or less nice solution I see, is to pass the height and width to `GetMap`. Would you prefer that @pblottiere ?

**Comment by pblottiere** [comment]
Hi @signedav, 

> If we don't decide on the REQUEST what HEIGHT is used there, it can be, that on a GetLegendGraphics request we compare the legend-image-size to the max-size of the map, what could lead to a wrong error message.

Arrghh... It's a shame to call the whole `getMap()` method when we just want to run a hit test :/. Whatever, considering the current implementation, I think your solution is the wiser for now!

Can you rename `getHeightAsInt()` -> `height()` and `getWidthAsInt()` -> `width()` please? Then I'll merge your PR as soon as Travis is happy.

Thanks for the discussion :).

**Comment by pblottiere** [comment]
Thanks @signedav!

**Comment by signedav** [comment]
Thank you @pblottiere as well.

**Pull request merged**